src/libsyntax/parse/parser/diagnostics.rs

```diff
@@ -916,6 +916,54 @@ impl<'a> Parser<'a> {
         Err(err)
     }
 
+    pub(super) fn expect_semi(&mut self) -> PResult<'a, ()> {
+        if self.eat(&token::Semi) {
+            return Ok(());
+        }
+        let sm = self.sess.source_map();
+        if self.look_ahead(1, |t| {
+            (t == &token::CloseDelim(token::Brace) || token_can_begin_expr(t)) && t.kind != token::Colon
```

Comment:
Two points here:
1. `t == token::CloseDelim(...)` implies `t.kind != token::Colon` so this check is redundant in the LHS of `||` and can be moved into the RHS.
2. It's easier to read if you use the rustfmt formatting with:
```rust
long_expression_seriously_this_is_long
binop some_other_random_expression
```